src/circuit.rs

```diff
@@ -485,30 +482,8 @@ impl plonk::Circuit<pallas::Base> for Circuit {
         let nf_old = {
             // hash_old = poseidon_hash(nk, rho_old)
             let hash_old = {
-                let message = [nk, rho_old];
-
-                let poseidon_message = layouter.assign_region(
-                    || "load message",
-                    |mut region| {
-                        let mut message_word = |i: usize| {
-                            let value = message[i].value();
-                            let var = region.assign_advice(
-                                || format!("load message_{}", i),
-                                config.poseidon_config.state[i],
-                                0,
-                                || value.ok_or(plonk::Error::SynthesisError),
-                            )?;
-                            region.constrain_equal(var, message[i].cell())?;
-                            Ok(Word::<_, _, poseidon::P128Pow5T3, 3, 2>::from_inner(
-                                StateWord::new(var, value),
-                            ))
-                        };
-
-                        Ok([message_word(0)?, message_word(1)?])
-                    },
-                )?;
```

Comment:
SAFETY: Removal of this region is fine:

- This region was copying the message pieces into new word cells.
- `PoseidonHash::hash` then passes those word cells to `Duplex::absorb`:
  https://github.com/zcash/orchard/blob/dfcea20569541399a32196c3cbb05af26c53afb1/src/circuit/gadget/poseidon.rs#L253-L257
- `Duplex` stores the word cells in an inner array, which is then passed to `poseidon_duplex`, and through to the `pad_and_add` instruction:
  https://github.com/zcash/orchard/blob/dfcea20569541399a32196c3cbb05af26c53afb1/src/circuit/gadget/poseidon.rs#L102-L104
- `pad_and_add` copies the word cells into a new region:
  https://github.com/zcash/orchard/blob/dfcea20569541399a32196c3cbb05af26c53afb1/src/circuit/gadget/poseidon/pow5t3.rs#L356-L381

So:
- Prior to this change, we were copying the origin cells into word cells, and then into the pad-and-add region.
- After this change, we copy the origin cells into the pad-and-add region directly.